dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/InstrumentedDnsResolver.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.hc5;
+
+import com.palantir.tracing.CloseableTracer;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import org.apache.hc.client5.http.DnsResolver;
+
+/** {@link DnsResolver} wrapper which adds tracing spans. */
+final class InstrumentedDnsResolver implements DnsResolver {
+
+    private final DnsResolver delegate;
+
+    InstrumentedDnsResolver(DnsResolver delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public InetAddress[] resolve(String host) throws UnknownHostException {
+        try (CloseableTracer ignored = CloseableTracer.startSpan("DnsResolver.resolve")) {
+            return delegate.resolve(host);
+        }
```

Comment:
I like having this tracing available to debug stuff - my concern is that the trace log viewer can already get really noisy when there are hundreds of RPC calls.

Would it be crazy to short-circuit this based on the log.isDebugEnabled? I'm thinking under 'normal' conditions, even when a trace is sampled, most people don't care about the DNS resolving.  If we're really trying to dig into a perf thing though we could set `com.palantir.dialogue.hc5: DEBUG` and then get these sweet granular spans??

Reply:
Actually scratch that the trace-log-viewer can easily be configured to hide spans that are very small, so I think we might as well just record this stuff. It looks incredibly fast in the unsampled case!